kubernetes/scheduler/controller.go

```diff
@@ -0,0 +1,668 @@
+				return
+			}
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(new)
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", new, err))
+				return
+			}
+			controller.workqueue.Add(key)
+
+		},
+		DeleteFunc: func(obj interface{}) {
+			node, ok := obj.(*corev1.Node)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("Couldn't cast object to node for object %#v", obj))
+				return
+			}
+			// Delete the node from the nodeResources map. Do it here, so it's
+			// not an item added to the queue.
+			delete(controller.nodeResources, nodeName(node.ObjectMeta.Name))
+		},
+	})
+
+	// Set up the map[pod.ObjectMeta.UID] errChan,  a unique channel for
+	// handling errors related to the passed pod. Uses the sync.Map as it fits
+	// use case 1 exactly: "The Map type is optimized for two common use cases:
+	// (1) when the entry for a given key is only ever written once but read
+	// many times.."
+	controller.podAliveMap = new(sync.Map)
+
+	return controller
+}
+
+// Run sets up event handlers, synces informer caches and starts workers. It
+// blocks until stopCh is closed, at which point it'll shut down workqueue and
+// wait for workers to finish processing.
+func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
+	c.Logger = c.opts.Logger.New("schedulerController", "kubernetes")
+	kubeLogFile := filepath.Join(c.opts.ManagerDir, kubeSchedulerControllerLog)
+	fh, err := log15.FileHandler(kubeLogFile, log15.LogfmtFormat())
+	if err != nil {
+		return fmt.Errorf("wr kubernetes scheduler could not log to %s: %s", kubeLogFile, err)
+	}
+
+	l15h.AddHandler(c.Logger, fh)
+	c.Debug("In Run()")
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	// Start informer factories, begin populating informer caches.
+
+	// Wait for caches to sync before starting workers
+	c.Debug("Waiting for caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, c.podSynced, c.nodeSynced); !ok {
+		c.Crit("failed to wait for caches to sync")
+		utilruntime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+	c.Debug("Caches synced")
+
+	c.Debug("In Run(), starting workers")
+	// start workers
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runReqCheck, time.Second, stopCh)
+		go wait.Until(c.runPodAlive, time.Second, stopCh)
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+
+	return nil
+}
+
+// continually call processNextWorkItem to read and process the next message on
+// the workqueue
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem reads a single work item off the workqueue and attempts
+// to process it, by calling the syncHandler.
+func (c *Controller) processNextWorkItem() bool {
+	obj, shutdown := c.workqueue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	// Wrap this block in func so can defer c.workqueue.Done.
```

Comment:
What's stopping you deferring the Done without the func wrapper?

Reply:
Done.